MediaRecorder.bs

```diff
@@ -410,6 +405,67 @@ interface MediaRecorder : EventTarget {
+    <li>Let |withTrack| be the method's second argument.</li>
+    <li>If |recorder|'s {{state}} is {{inactive}}, return a promise rejected
+    with a newly created {{InvalidStateError}} {{DOMException}} and abort these
+    steps.</li>
+    <li>If |track| is not found in |recorder|'s [=[[Tracks]]=], return a promise
+    rejected with a newly created {{NotFoundError}} {{DOMException}} and abort
+    these steps.</li>
+    <li>If |track| is {{ended}}, or |withTrack|'s {{kind}} differs from
+    |track|'s {{kind}}, then return a promise rejected with a newly created
+    {{InvalidModificationError}} {{DOMException}} and abort these steps.</li>
+    <li>If the <a>isolation properties</a> of |track| or |withTrack| disallow
+    access from |recorder|, return a promise rejected with a newly created
+    {{SecurityError}} {{DOMException}} and abort these steps.</li>
+    <li>If the User Agent cannot record |withTrack| in place of |track| using
+    the current configuration, then return a promise rejected with a newly
+    created {{NotSupportedError}} {{DOMException}} and abort these steps.</li>
+    <li>Let |session| be the ongoing recording session corresponding to the
+    first occurance of |track| in |recorder|'s [=[[Tracks]]=].</li>
+    <li>Replace the first occurance of |track| in |recorder|'s [=[[Tracks]]=]
+    with |withTrack|.</li>
+    <li>Add |session| and |withTrack| as a pair to the list in |recorder|'s
+    [=[[ReplaceAtOnceList]]=] slot.</li>
+    <li>If |recorder|'s [=[[ReplacePromise]]=] slot is not <code>null</code>,
+    return it, and abort these steps.</li>
+    <li>Let |p| be a new promise.</li>
+    <li>Set |recorder|'s [=[[ReplacePromise]]=] slot to |p|, and queue
+    a task, using the DOM manipulation task source, that runs the following
+    steps:</li>
+    <ol>
+      <li>If |recorder|'s {{state}} attribute is {{inactive}}, resolve |p|, and
+      abort these steps.</li>
+      <li>Let |replaceList| be |recorder|'s [=[[ReplaceAtOnceList]]=], and set
+      |recorder|'s [=[[ReplaceAtOnceList]]=] to an empty list.</li>
+      <li>Set |recorder|'s [=[[ReplacePromise]]=] slot to <code>null</code>, and
+      run the remaining steps in parallel:</li>
+      <li>For each recording |session| and |withTrack| pair in |replaceList|,
+      instantly switch |session| to recording |withTrack| in place of what
+      |session| is currently recording.
+      <div class="note">This replaces the source of frames being gathered and
+      SHOULD be a seamless operation. This SHOULD happen at the same time for
+      all sessions in |recorder|'s [=[[ReplaceAtOnceList]]=].</div>
+      </li>
+      <li>Resolve |p|.
+      <div class="note">|p| MUST NOT be rejected for any reason, since track
+      replacement is observably synchronous in the API by calling |replaceTrack|
+      multiple times. The promise exists to simplify learning when replaced
+      tracks can be stopped without impacting the recording.</div></li>
```

Comment:
Can you give an example of when `recorder.replaceTrack(track, withTrack); track.stop();` would impact the recording so that `track` is not replaced with `withTrack`, given that all the attributes checked synchronously are OK?

I.e., explain more in detail why the promise is needed, please.

Reply:
@henbos I believe you raised the need for a promise. Care to comment?

The only case I can think of is bad behaving JS, e.g.:
```js
recorder.replaceTrack(videoA, newVideoA);
videoA.stop();
for (const i = 0; i < 1000000; i++) x += i / 3.14;
recorder.replaceTrack(audioB, newAudioB);
audioA.stop();
```
In this case would audio and video switch at the same time?

Reply:
Yes, but the recording would also stop immediately after they are replaced, given that the recorder is only recording `videoA` and `audioA`.

According to this PR, the first `replaceTrack` queues task A to atomically replace both `videoA` and `audioA`. `audioA.stop()` makes "all tracks become ended" (from the spec) and queues a task B to inactivate the recorder and fire "stop". Task A runs before task B, but after `audioA.stop()` so there's your answer.

Having a promise doesn't stop task B from inactivating the recorder. I still don't see the need for the promise.

I do however see a need for having task B per above exit early (the `Tracks` slot contains live tracks!), or avoid being fired (the `ReplaceAtOnceList` slot is not empty!).